Runtime/C++/src/bebop.hpp

```diff
@@ -0,0 +1,292 @@
+using TickDuration = std::chrono::duration<int64_t, Tick>;
+
+namespace {
+    /// The number of ticks between 1/1/0001 and 1/1/1970.
+    const int64_t ticksBetweenEpochs = 621355968000000000;
+}
+
+enum class GuidStyle {
+    Dashes,
+    NoDashes,
+};
+
+struct Guid {
+    /// The GUID bytes, in .NET "ToByteArray" order.
+    /// https://docs.microsoft.com/en-us/dotnet/api/system.guid.tobytearray?view=net-5.0
+    uint8_t m_bytes[16];
+
+    Guid() = default;
+    Guid(const uint8_t *bytes) { std::memcpy(m_bytes, bytes, 16); }
+    Guid(Guid const &other) : Guid(other.m_bytes) {}
+
+    static Guid fromString(const std::string &string) {
+        Guid g;
+        static const uint8_t asciiToHex[256] = {
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  0,  0,  0,  0,  0,  0,
+            0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+            0, 10, 11, 12, 13, 14, 15,  // and the rest is zeroes
+        };
+        const char* s = string.c_str(); uint8_t a;
+
+#define Nibble() (asciiToHex[static_cast<unsigned char>(*s++)])
+#define B(i) a = Nibble(); a = (a << 4) | Nibble(); g.m_bytes[i] = a;
+#define Dash if (*s == '-') s++;
+        B(3)B(2)B(1)B(0) Dash B(5)B(4) Dash B(7)B(6) Dash B(8)B(9) Dash B(10)B(11)B(12)B(13)B(14)B(15)
+#undef Dash
+#undef B
+#undef Nibble
+
+        return g;
+    }
+
+    std::string toString(GuidStyle style = GuidStyle::Dashes) {
+        static const char nibbleToHex[16] = {'0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'};
+        std::string result;
+        bool dashes = style == GuidStyle::Dashes;
+        result.reserve(36);
+        uint8_t a;
+#define B(i) a = m_bytes[i]; result += nibbleToHex[a >> 4]; result += nibbleToHex[a & 0xf];
+#define Dash if (dashes) result += '-';
+        B(3)B(2)B(1)B(0) Dash B(5)B(4) Dash B(7)B(6) Dash B(8)B(9) Dash B(10)B(11)B(12)B(13)B(14)B(15)
+#undef Dash
+#undef B
+        return result;
+    }
+
+    bool operator<(const Guid &other) const {
+        for (size_t i = 0; i < 16; i++) {
+            if (m_bytes[i] < other.m_bytes[i]) return true;
+            if (m_bytes[i] > other.m_bytes[i]) return false;
+        }
+        return false;
+    }
+
+    bool operator==(const Guid &other) const {
+        for (size_t i = 0; i < 16; i++) {
+            if (m_bytes[i] != other.m_bytes[i]) return false;
+        }
+        return true;
+    }
+};
+
+class BebopReader {
+    const uint8_t *m_pointer;
+    BebopReader() {}
+public:
+    BebopReader(BebopReader const&) = delete;
+    void operator=(BebopReader const&) = delete;
+    static BebopReader& instance(const uint8_t *buffer) {
+        static BebopReader instance;
```

Comment:
Yea I think your comment about not needing this to be a Singleton is probably a good one. I'd just have this class be a regular class.